test/index.js

```diff
@@ -1589,6 +1589,33 @@ describe('def', function() {
                    '  - (Either ??? ???)\n'));
   });
 
+  it('supports values of "foreign" types', function() {
+    //  id :: a -> a
+    var id = def('id', {}, [a, a], R.identity);
+
+    eq(id({'@@type': 'my-package/Foreign'}), {'@@type': 'my-package/Foreign'});
+
+    throws(function() { id([{'@@type': 'my-package/Foo'},
+                            {'@@type': 'my-package/Bar'}]); },
+           errorEq(TypeError,
+                   '‘id’ received [{"@@type": "my-package/Foo"}, ' +
+                   '{"@@type": "my-package/Bar"}] as its first argument, ' +
```

Comment:
This turned out to be much more involved than I anticipated, but I'm very pleased with the results. Have a look at the changes to the test suite, @soroushjp. :)